processor/memorylimiter/memorylimiter.go

```diff
@@ -238,44 +245,93 @@ func (ml *memoryLimiter) forcingDrop() bool {
 	return atomic.LoadInt64(&ml.forceDrop) != 0
 }
 
-func (ml *memoryLimiter) memCheck() {
+func (ml *memoryLimiter) setForcingDrop(b bool) {
+	var i int64
+	if b {
+		i = 1
+	}
+	atomic.StoreInt64(&ml.forceDrop, i)
+}
+
+func memstatToZapField(ms *runtime.MemStats) zap.Field {
+	return zap.Uint64("cur_mem_mib", ms.Alloc/1024/1024)
+}
+
+func (ml *memoryLimiter) doGCandReadMemStats() *runtime.MemStats {
+	runtime.GC()
+	ml.lastGCDone = time.Now()
 	ms := ml.readMemStats()
-	ml.memLimiting(ms)
+	ml.logger.Info("Memory usage after GC.", memstatToZapField(ms))
+	return ms
 }
 
-func (ml *memoryLimiter) memLimiting(ms *runtime.MemStats) {
-	if !ml.decision.shouldDrop(ms) {
-		atomic.StoreInt64(&ml.forceDrop, 0)
-	} else {
-		atomic.StoreInt64(&ml.forceDrop, 1)
-		// Force a GC at this point and see if this is enough to get to
-		// the desired level.
-		runtime.GC()
+func (ml *memoryLimiter) checkMemLimits() {
+	ms := ml.readMemStats()
+
+	ml.logger.Debug("Currently used memory.", memstatToZapField(ms))
+
+	if ml.usageChecker.aboveHardLimit(ms) {
+		ml.logger.Warn("Memory usage is above hard limit. Forcing a GC.", memstatToZapField(ms))
+		ms = ml.doGCandReadMemStats()
+	}
+
+	// Remember current dropping state.
+	wasForcingDrop := ml.forcingDrop()
+
+	// Check if the memory usage is above the soft limit.
+	mustForceDrop := ml.usageChecker.aboveSoftLimit(ms)
+
+	if wasForcingDrop && !mustForceDrop {
+		// Was previously dropping but enough memory is available now, no need to limit.
+		ml.logger.Info("Memory usage back within limits. Resuming normal operation.", memstatToZapField(ms))
+	} else if !wasForcingDrop && mustForceDrop {
+		// We are above soft limit, do a GC if it wasn't done recently and see if
+		// it brings memory usage below the soft limit.
+		if time.Since(ml.lastGCDone) > minGCIntervalWhenSoftLimited {
```

Comment:
Other than potentially calling `ml.setForcingDrop(mustForceDrop)`, are you doing anything in this function when the last GC happened less than 10s ago? If not, you might want to move this check to the top of the function, and return early.

Reply:
>are you doing anything in this function when the last GC happened less than 10s ago?

Yes, we may still do the GC if we are above hard limit. We also need to check the soft limit and change the drop flag, so I don't see how we can return early only by checking this condition.